### PR TITLE
MAINT: Allow subclasses in `ndarray.__array_function__`.

### PR DESCRIPTION
**Pull request by mhvk**
The Liskov substitution principle suggests it should.

### CODE REVIEW

File: numpy/core/overrides.py [hunk 1]
```diff
@@ -58,16 +58,6 @@ def get_overloaded_types_and_args(relevant_args):
                 overloaded_types = [arg_type]
                 overloaded_args = [arg]
 
-    # Short-cut for the common case of only ndarray.
```
Comment (shoyer) [inline numpy/core/overrides.py]: Just to be sure I understand: this (minor) optimization is enabled by your `ndarray.__array_function__` refactor, but isn't required?

Comment (mhvk) [reply numpy/core/overrides.py]: Indeed, I moved it to the actual execution part as it didn't make sense to check whether there are any non-ndarray overrides here.

**Comment by mhvk** [comment]
OK, I now omit all tests for the override for the standard case - since these are still run at least once, I think that is fine (but can make it more specific if you think it is warranted).

**Comment by mhvk** [comment]
OK, finally got the right ones all skipped if no `__array_ufunc__` is used.

**Comment by shoyer** [comment]
thanks @mhvk !

**Pull request merged**